### PR TITLE
Allow multiple storage classes

### PR DESCRIPTION
**Pull request by heytrav**
Rebase and small refactoring of #32

### CODE REVIEW

File: charts/cluster-addons/templates/openstack/csi-cinder.yaml [hunk 1]
```diff
@@ -72,27 +72,55 @@ spec:
...
           {{- with .volumeType }}
           type: {{ . }}
           {{- end }}
+          {{- with .fstype }}
+          fstype: {{ . }}
+          {{- end }}
         reclaimPolicy: {{ .reclaimPolicy }}
         allowVolumeExpansion: {{ .allowVolumeExpansion }}
-        volumeBindingMode: WaitForFirstConsumer
+        volumeBindingMode: {{ .volumeBindingModea }}
```
Comment (mkjpryor) [inline charts/cluster-addons/templates/openstack/csi-cinder.yaml:93]: Typo

File: charts/cluster-addons/templates/openstack/csi-cinder.yaml [hunk 2]
```diff
@@ -72,27 +72,55 @@ spec:
...
         {{- with .allowedTopologies }}
         allowedTopologies: {{ toYaml . | nindent 6 }}
         {{- end }}
         {{- end }}
+{{- $storeageClassIndex := 0 }}
+{{- range .Values.openstack.csiCinder.additionalStorageClasses }}
+{{- $rangeItem := . -}}
```
Comment (mkjpryor) [inline charts/cluster-addons/templates/openstack/csi-cinder.yaml:100]: I think you can do this as

```
{{- range $rangeItem := .Values.openstack.csiCinder.additionalStorageClasses }}
```

??

File: charts/cluster-addons/templates/openstack/csi-cinder.yaml [hunk 3]
```diff
@@ -72,27 +72,55 @@ spec:
...
+          {{- end }}
         reclaimPolicy: {{ .reclaimPolicy }}
         allowVolumeExpansion: {{ .allowVolumeExpansion }}
-        volumeBindingMode: WaitForFirstConsumer
+        volumeBindingMode: {{ .volumeBindingModea }}
         {{- with .allowedTopologies }}
         allowedTopologies: {{ toYaml . | nindent 6 }}
         {{- end }}
         {{- end }}
+{{- $storeageClassIndex := 0 }}
```
Comment (mkjpryor) [inline charts/cluster-addons/templates/openstack/csi-cinder.yaml:98]: Typo

Also, what is this actually used for? As far as I can see it is incremented but never used.

File: charts/cluster-addons/templates/openstack/csi-cinder.yaml [hunk 4]
```diff
@@ -72,27 +72,55 @@ spec:
...
+{{- range .Values.openstack.csiCinder.additionalStorageClasses }}
+{{- $rangeItem := . -}}
+{{- with $ }}
```
Comment (mkjpryor) [inline charts/cluster-addons/templates/openstack/csi-cinder.yaml:101]: Is this actually necessary? I don't see any use of `.Values`, `.Chart`, etc. inside the loop?

File: charts/cluster-addons/values.yaml [hunk 5]
```diff
@@ -89,25 +88,32 @@ openstack:
...
+      # reclaimPolicy: Delete
+      # # Indicates if volume expansion is allowed
+      # allowVolumeExpansion: true
+      # # The Cinder availability zone to use for volumes provisioned by the storage class
+      # availabilityZone: nova
+      # # The Cinder volume type to use for volumes provisioned by the storage class
+      # # If not given, the default volume type will be used
+      # volumeType: fast
+      # # The allowed topologies for the storage class
+      # allowedTopologies:
```
Comment (mkjpryor) [inline charts/cluster-addons/values.yaml:103]: Please keep the same defaults as we had previously so as not to break existing installations that just use the defaults.

File: charts/cluster-addons/values.yaml [hunk 6]
```diff
@@ -89,25 +88,32 @@ openstack:
...
-    storageClass:
-      # Indicates if the storage class should be enabled
-      enabled: true
-      # The name of the storage class
-      name: csi-cinder
-      # Indicates if the storage class should be annotated as the default storage class
-      isDefault: true
-      # The reclaim policy for the storage class
+    defaultStorageClass:
+      name: csi-cinder-fast
```
Comment (mkjpryor) [inline charts/cluster-addons/values.yaml:92]: This should be just `csi-cinder` to match the old default

File: charts/cluster-addons/values.yaml [hunk 7]
```diff
@@ -89,25 +88,32 @@ openstack:
...
-      volumeType:
-      # The allowed topologies for the storage class
+      # # The Cinder volume type to use for volumes provisioned by the storage class
+      # # If not given, the default volume type will be used
+      volumeType: fast
+      # # The allowed topologies for the storage class
       allowedTopologies:
+    additionalStorageClasses: []
+      # - name: csi-cinder-fast
+      #   # The reclaim policy for the storage class
```
Comment (mkjpryor) [inline charts/cluster-addons/values.yaml:106]: This should document `volumeBindingMode` as being required for additional storage classes

**Review by mkjpryor — CHANGES_REQUESTED** [review]
Sorry @heytrav - just a few more things that I missed the first time

File: charts/cluster-addons/values.yaml [hunk 8]
```diff
@@ -8,7 +7,7 @@ clusterName:
 hooks:
   image:
     repository: ghcr.io/stackhpc/capi-helm-utils
-    tag:  # Defaults to chart appVersion if not given
+    tag: # Defaults to chart appVersion if not given
```
Comment (mkjpryor) [inline charts/cluster-addons/values.yaml:10]: This space is important for the linting

**Review by mkjpryor — CHANGES_REQUESTED** [review]

**Comment by mkjpryor** [comment]
@heytrav Some changes were made to the cloud where we run the CI. Please can you rebase to pick up the required config changes?

**Review by mkjpryor — APPROVED** [review]
LGTM

**Comment by mkjpryor** [comment]
@heytrav 

The chart upgrade test is failing with the following error:

```
Error: UPGRADE FAILED: cannot patch "csi-cinder" with kind StorageClass: StorageClass.storage.k8s.io "csi-cinder" is invalid: parameters: Forbidden: updates to parameters are forbidden.
```

Looking at the patch, I think this is due to attempting to add `fstype: ext4` to the storage class. To be consistent with the previous setup, we need to not set this for the default storage class. I think the default is `ext4` anyway?

**Comment by mkjpryor** [comment]
I feel like using the Kubernetes default fstype by default is the desired behaviour anyway, rather than the chart having an explicit opinion about it being `ext4`.

File: charts/cluster-addons/values.yaml [hunk 9]
```diff
@@ -84,8 +79,30 @@ openstack:
       # The Cinder volume type to use for volumes provisioned by the storage class
       # If not given, the default volume type will be used
       volumeType:
+      # The volumeBindingMode field controls when volume binding and dynamic provisioning should occur.
+      volumeBindingMode: WaitForFirstConsumer
       # The allowed topologies for the storage class
       allowedTopologies:
+      # Filesystem type supported by Kubernetes. Defaults to "ext4"
+      fstype: ext4
```
Comment (mkjpryor) [inline charts/cluster-addons/values.yaml:87]: Do not set a default here.

**Review by mkjpryor — CHANGES_REQUESTED** [review]

File: charts/cluster-addons/values.yaml (hunk 9 above)
Comment (heytrav) [reply charts/cluster-addons/values.yaml]: Ok I removed ext4 as the default

**Comment by mkjpryor** [comment]
@heytrav Did you mean to do all that other reformatting of the values file?

**Comment by heytrav** [comment]
> @heytrav Did you mean to do all that other reformatting of the values file?

I didn't mean to no. My editor's linter settings are a bit too opinionated.

**Review by mkjpryor — APPROVED** [review]
LGTM subject to CI

**Pull request merged**